### PR TITLE
[MID-155] Navigation links (admin panel + ui)

### PR DESCRIPTION
**Pull request by watchlar2000**
[MID-155](https://trello.com/c/ClDiKEdE/155-%D0%B4%D0%BE%D0%B4%D0%B0%D1%82%D0%B8-%D0%B2-%D0%B0%D0%B4%D0%BC%D1%96%D0%BD%D0%BA%D1%83-%D0%BC%D0%BE%D0%B6%D0%BB%D0%B8%D0%B2%D1%96%D1%81%D1%82%D1%8C-%D0%B7%D0%BC%D1%96%D0%BD%D0%B8%D1%82%D0%B8-%D0%BB%D1%96%D0%BD%D0%BA%D0%B8-%D0%BD%D0%B0-%D0%B7%D0%B0%D0%BB%D0%B8%D1%88%D0%B8%D1%82%D0%B8-%D0%B7%D0%B0%D1%8F%D0%B2%D0%BA%D1%83-%D1%82%D0%B0-%D1%81%D0%BE%D1%86%D1%96%D0%B0%D0%BB%D1%8C%D0%BD%D1%96-%D0%BC%D0%B5%D1%80%D0%B5%D0%B6%D1%96-%D1%8F%D0%BA%D1%96-%D1%80%D0%BE%D0%B7%D1%82%D0%B0%D1%88%D0%BE%D0%B2%D0%B0%D0%BD%D1%96-%D0%B2-header-%D1%82%D0%B0-footer)

Admin now can edit media links though admin panel.

### CODE REVIEW

File: src/app/_components/Footer.js [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
+import { getSocialMediaList } from '@components/Links/socialMediaActions';
 
 export function Footer() {
-  const { links } = siteNav;
+  const [applicationLink, setApplicationLink] = useState('');
+
+  useEffect(() => {
+    getSocialMediaList().then(data => {
+      setApplicationLink(data.applicationLink);
+    });
```
Comment (ZakharovMaxim) [inline src/app/_components/Footer.js]: Can we transform footer.js to a server component and get the links before we render?

File: src/app/_components/Links/SocialLinksList.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import { getSocialMediaList } from '@components/Links/socialMediaActions';
+
+export function SocialLinksList({ className }) {
+  const [socialMediaList, setSocialMediaList] = useState(null);
+
+  useEffect(() => {
+    getSocialMediaList().then(data => {
+      setSocialMediaList(data.socialMediaList);
+    });
+  }, []);
```
Comment (ZakharovMaxim) [inline src/app/_components/Links/SocialLinksList.js]: Transform this component to a server component

File: src/app/_components/Links/socialMediaActions.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+export async function getSocialMediaList() {
+  const data = await prisma.navigation.findMany({
+    select: {
+      title: true,
+      href: true,
+    },
+  });
+
+  revalidateTag('/');
```
Comment (ZakharovMaxim) [inline src/app/_components/Links/socialMediaActions.js]: Why do we revalidate the homepage every time we request social links?

File: src/app/_components/Links/socialMediaActions.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  const { INSTAGRAM, APPLICATION } = NavigationUrl;
+
+  const socialMediaList = data
+    ?.filter(({ title }) => title !== APPLICATION)
+    .map(({ title, href }) => ({
+      title,
+      href,
+      svg: title === INSTAGRAM ? <Instagram /> : <Facebook />,
```
Comment (ZakharovMaxim) [inline src/app/_components/Links/socialMediaActions.js]: Do it on the client side instead, e.g. in SocialLinksList.js

**Review by ZakharovMaxim — CHANGES_REQUESTED** [review]

File: src/app/_components/Footer.js (hunk 1 above)
Comment (watchlar2000) [reply src/app/_components/Footer.js]: Yeah, we can do that, thanks!

File: src/app/_components/Links/SocialLinksList.js (hunk 2 above)
Comment (watchlar2000) [reply src/app/_components/Links/SocialLinksList.js]: Component transformed to a server one

File: src/app/_components/Links/socialMediaActions.js (hunk 4 above)
Comment (watchlar2000) [reply src/app/_components/Links/socialMediaActions.js]: Done. Thanks!

File: src/app/_components/Links/socialMediaActions.js (hunk 3 above)
Comment (watchlar2000) [reply src/app/_components/Links/socialMediaActions.js]: Let's just keep it as a temporary solution for now and will get back to it later, ok?

File: src/app/_hooks/useBodyScrollLock.js [hunk 5]
```diff
@@ -1,25 +1,31 @@
...
+
+export function useBodyScrollLock(locked) {
+  // const axes = useMemo(
+  //   () => ({
+  //     x: 'overflow-x-hidden',
+  //     y: 'overflow-y-hidden',
+  //     xy: 'overflow-hidden',
+  //   }),
+  //   [],
+  // );
```
Comment (ZakharovMaxim) [inline src/app/_hooks/useBodyScrollLock.js]: Uncomment it and use, but you can move it outside of the hook, no need to wrap it in useMemo

File: src/app/_components/Modal/Modal.js [hunk 6]
```diff
@@ -59,7 +63,7 @@ export const Modal = ({
                   <ModalCloseButton onClose={onClose} />
                 </div>
               )}
-              {children}
+              <div className={cn('mt-4 overflow-y-auto')}>{children}</div>
```
Comment (ZakharovMaxim) [inline src/app/_components/Modal/Modal.js]: cn is not needed here

File: src/app/_components/Links/linksActions.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    select: {
+      title: true,
+      href: true,
+    },
+  });
+
+  const { APPLICATION } = NavigationUrl;
+
+  // temporary solution to ensure the links are up-to-date with the db
+  revalidateTag('/');
```
Comment (ZakharovMaxim) [inline src/app/_components/Links/linksActions.js]: Remove this for now

File: src/app/_components/EventLinkModal.js [hunk 8]
```diff
@@ -47,6 +47,7 @@ export function EventLinkModal({ isOpen, onClose, link }) {
       className="w-64 opacity-100 lg:w-fit lg:bg-primary-200"
       isCloseButton={isDesktop}
       layout={isDesktop}
+      scrollY
```
Comment (ZakharovMaxim) [inline src/app/_components/EventLinkModal.js]: Do we have this prop?

File: src/app/(app)/layout.js [hunk 9]
```diff
@@ -16,7 +17,7 @@ export default function Layout({ children }) {
   return (
     <Hint>
       <div className="flex min-h-screen flex-col">
-        <Header />
+        <Header socialLinks={<SocialLinksList className="text-primary-700 hover:text-primary-500" />} />
```
Comment (ZakharovMaxim) [inline src/app/(app)/layout.js]: Can we get social links here and pass them as a prop to the header and footer?

File: src/app/_components/Links/SocialLinksList.js [hunk 10]
```diff
@@ -1,7 +1,18 @@
 import Link from 'next/link';
 import PropTypes from 'prop-types';
+import { getLinksList } from '@components/Links/linksActions';
+import { NavigationUrl } from '@prisma/client';
+import { Facebook, Instagram } from '@icons';
+
+export async function SocialLinksList({ className }) {
+  const { socialMediaList } = await getLinksList();
```
Comment (ZakharovMaxim) [inline src/app/_components/Links/SocialLinksList.js]: If you move this call to the layout, then we can remove it here

File: src/app/_components/Header.js [hunk 11]
```diff
@@ -38,7 +38,9 @@ export function Header() {
     setDonateModalOpen(prevState => !prevState);
   }, [setDonateModalOpen]);
 
-  useBodyScrollLock(isMenuOpen, 'y');
+  const isScrollLock = isMenuOpen || isFeedbackOpen || isDonateModalOpen;
+  useBodyScrollLock(isScrollLock);
```
Comment (ZakharovMaxim) [inline src/app/_components/Header.js]: I don't really like it, I prefer when this logic is inside a modal/drawer component, so no matter where do we render them they'll behave the same

**Review by ZakharovMaxim — CHANGES_REQUESTED** [review]

File: src/app/_hooks/useBodyScrollLock.js (hunk 5 above)
Comment (watchlar2000) [reply src/app/_hooks/useBodyScrollLock.js]: It looks like this useMemo should be there inside the hook. What do you think?

image

File: src/app/_components/Modal/Modal.js (hunk 6 above)
Comment (watchlar2000) [reply src/app/_components/Modal/Modal.js]: Sure, thanks!

File: src/app/_components/Links/linksActions.js (hunk 7 above)
Comment (watchlar2000) [reply src/app/_components/Links/linksActions.js]: Removed

File: src/app/_components/EventLinkModal.js (hunk 8 above)
Comment (watchlar2000) [reply src/app/_components/EventLinkModal.js]: It was there, but eventually I came up with another idea) thanks for noticing it!

File: src/app/(app)/layout.js (hunk 9 above)
Comment (watchlar2000) [reply src/app/(app)/layout.js]: I think we can do that. Though, how should "layout" page be revalidated?

File: src/app/_components/Links/SocialLinksList.js (hunk 10 above)
Comment (watchlar2000) [reply src/app/_components/Links/SocialLinksList.js]: Have done it!

**Review by BigTako — APPROVED** [review]

File: src/app/_components/Header.js (hunk 11 above)
Comment (watchlar2000) [reply src/app/_components/Header.js]: Finally figured out how to do it, thanks! Now it should look better

**Review by alex-popov-tech — APPROVED** [review]

File: src/app/_components/Event/Card.js [hunk 12]
```diff
@@ -44,6 +46,11 @@ function transformData(event) {
 
 export function EventCard({ event }) {
   const [isModalOpenOpen, setIsModalOpen] = useState(false);
+
+  useKeyPress(KEY_TO_CLOSE_MODAL, () => {
+    setIsModalOpen(false);
+  });
```
Comment (ZakharovMaxim) [inline src/app/_components/Event/Card.js]: Can we move it to the modal component itself?

File: src/app/_components/Footer.js [hunk 13]
```diff
@@ -1,12 +1,12 @@
...
 import { cn } from '@utils/cn';
-import siteNav from '@config/siteNav';
 import { PillButton } from '@components/PillButton';
-import { SocialLink } from '@components/Links';
+import PropTypes from 'prop-types';
+import { linkItemPropType, SocialLinksList } from '@components/Links';
 
-export function Footer() {
-  const { links } = siteNav;
+export async function Footer({ socialLinks }) {
```
Comment (ZakharovMaxim) [inline src/app/_components/Footer.js]: It should not be async anymore

File: src/app/_components/CardSpecialist/CardModalWrapper.js [hunk 14]
```diff
@@ -20,14 +22,14 @@ export function CardModalWrapper({ children, className }) {
     setIsOpen(true);
   }, []);
 
-  useBodyScrollLock(isOpen, 'y');
+  useKeyPress(KEY_TO_CLOSE_MODAL, handleClose);
```
Comment (ZakharovMaxim) [inline src/app/_components/CardSpecialist/CardModalWrapper.js]: Move it to the Modal component

File: src/app/_hooks/useKeyPress.js [hunk 15]
```diff
@@ -10,9 +10,9 @@ export function useKeyPress(key, handler) {
     [key, handler],
   );
   useEffect(() => {
-    window.addEventListener('keypress', onKeyPress);
+    window.addEventListener('keyup', onKeyPress);
```
Comment (ZakharovMaxim) [inline src/app/_hooks/useKeyPress.js]: I dont mind this change, but the hook's name is useKeyPress, is there a reason to change to keyup?

File: src/app/_components/Footer.js (hunk 13 above)
Comment (watchlar2000) [reply src/app/_components/Footer.js]: Sure

File: src/app/_components/Event/Card.js (hunk 12 above)
Comment (watchlar2000) [reply src/app/_components/Event/Card.js]: Sure!

File: src/app/_components/CardSpecialist/CardModalWrapper.js (hunk 14 above)
Comment (watchlar2000) [reply src/app/_components/CardSpecialist/CardModalWrapper.js]: Done

File: src/app/_hooks/useKeyPress.js (hunk 15 above)
Comment (watchlar2000) [reply src/app/_hooks/useKeyPress.js]: Changed it back to 'keypress'

File: src/app/globals.css [hunk 16]
```diff
@@ -2,57 +2,58 @@
 @tailwind components;
 @tailwind utilities;
 
+body {
```
Comment (ZakharovMaxim) [inline src/app/globals.css]: It should be html

Comment (watchlar2000) [reply src/app/globals.css]: Fixed

**Review by ZakharovMaxim — APPROVED** [review]

**Pull request merged**